cloud/scope/principal.go

```diff
@@ -0,0 +1,167 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package scope
+
+import (
+	"context"
+	"encoding/base64"
+	"fmt"
+
+	"github.com/Azure/go-autorest/autorest"
+	"github.com/Azure/go-autorest/autorest/adal"
+	"github.com/Azure/go-autorest/autorest/azure/auth"
+	"github.com/pkg/errors"
+
+	aadpodid "github.com/Azure/aad-pod-identity/pkg/apis/aadpodidentity"
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+
+	infrav1 "sigs.k8s.io/cluster-api-provider-azure/api/v1alpha3"
+	client "sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// type AzureIdentityProvider interface {
+// 	Authorizer
+
+// 	// Hash returns a unique hash of the data forming the credentials
+// 	// for this principal
+// 	Hash() (string, error)
+// }
+
+// AzureCredentialsProvider provides
+type AzureCredentialsProvider struct {
+	Client                  client.Client
+	Principal               *infrav1.AzureClusterPrincipal
+	PrincipalRef            *infrav1.AzurePrincipalRef
+	ResourceManagerEndpoint string
+	TenantID                string
+	ClientID                string
+	ClientSecret            string
+}
+
+// NewAzureCredentialsProvider creates a new AzureCredentialsProvider from the supplied inputs.
+func NewAzureCredentialsProvider(ctx context.Context, kubeclient client.Client, resourceManagerEndpoint string, ref *infrav1.AzurePrincipalRef) (*AzureCredentialsProvider, error) {
+	principal := infrav1.AzureClusterPrincipal{}
+	if ref == nil {
+		return nil, errors.New("failed to generate new AzureCredentialsProvider from nil PrincipalRef")
+	}
+	err := kubeclient.Get(ctx, client.ObjectKey{Name: ref.Name, Namespace: ref.Namespace}, &principal)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to get AzureClusterPrincipal")
+	}
+
+	secret := &corev1.Secret{}
+	key := types.NamespacedName{Name: principal.Spec.SecretRef.Name, Namespace: principal.Spec.SecretRef.Namespace}
+	if err := kubeclient.Get(ctx, key, secret); err != nil {
+		return nil, errors.Wrapf(err, "failed to retrieve bootstrap data secret for AzureMachine %s", key)
+	}
+	tenantID, err := getValueFromSecret(secret, "tenantID")
+	if err != nil {
+		return nil, err
+	}
+	clientID, _ := getValueFromSecret(secret, "clientID")
+	clientSecret, _ := getValueFromSecret(secret, "clientSecret")
+
+	return &AzureCredentialsProvider{
+		Client:                  kubeclient,
+		Principal:               &principal,
+		PrincipalRef:            ref,
+		ResourceManagerEndpoint: resourceManagerEndpoint,
+		TenantID:                tenantID,
+		ClientID:                clientID,
+		ClientSecret:            clientSecret,
+	}, nil
+}
+
+// GetAuthorizer returnes Azure authorizer based on the provided principal
+func (p *AzureCredentialsProvider) GetAuthorizer(ctx context.Context) (autorest.Authorizer, error) {
```

Comment:
Creating the Authorizer should be trivial using AAD pod identity. We can use the MSI endpoint provided. See the demo code here for an example of creating the Authorizer.

https://github.com/Azure/aad-pod-identity/blob/fb1bdba48ab5c1fa0be64382b6a74acdbcb41089/cmd/demo/main.go#L48